fixes.py

```diff
@@ -0,0 +1,253 @@
+
+        self.item = item
+        self.item_report = item_report
+
+
+    def tags_or_title_fix(self):
+        '''
+        Use item.update() to update title and/or tags.
+
+        Updates item_report with results for this fix:
+        {tags_title_result: result}
+        '''
+
+        title = self.item_report['title_new']
+        tags = self.item_report['tags_new']
+
+        if title or tags:
+
+            #: Tags and title combined .update()
+            update_dict = {}
+            messages = []
+            if title:
+                update_dict['title'] = title
+                messages.append(f'title to "{title}"')
+            if tags:
+                update_dict['tags'] = tags
+                messages.append(f'tags to {tags}')
+            update_result = self.item.update(update_dict)
+            if update_result:
+                result = f'Updated {", ".join(messages)}'
+            else:
+                result = f'Failed to update {", ".join(messages)}'
+        
+        else:
+            result = 'No update needed for title or tags'
+
+        self.item_report['tags_title_result'] = result
+
+
+    def group_fix(self, groups_dict):
+        '''
+        Use item.share() to share to group
+        groups_dict:    A dictionary of all the org's groups and their
+                        corresponding ids.
+
+        Updates item_report with results for this fix:
+        {groups_result: result}
+        '''
+
+        if self.item_report['groups_fix'] == 'Y':
+
+            group_name = self.item_report['group_new']
+
+            try:
+                gid = groups_dict[group_name]
+
+                share_results = self.item.share(everyone=True, groups=[gid])
+                result_dict = share_results['results'][0]
+                if gid not in result_dict['notSharedWith']:
+                    result = f'Shared with everyone and "{group_name}" group'
+                else:
+                    result = f'Failed to share with everyone and "{group_name}" group'
+
+            except KeyError:
+                result = f'Cannot find group "{group_name}" in organization'
+
+        else:
+            result = 'No update needed for groups'
+
+        self.item_report['groups_result'] = result
+
+
+    def folder_fix(self):
+        '''
+        Use item.move() to move item to folder.
+
+        Updates item_report with results for this fix:
+        {folder_result: result}
+        '''
+
+        if self.item_report['folder_fix'] == 'Y':
+            folder = self.item_report['folder_new']
+
+            move_result = self.item.move(folder)
+            if move_result:
+                if move_result['success']:
+                    result = f'Item moved to "{folder}" folder'
+                else:
+                    result = f'Failed to move item to "{folder}" folder'
+            else:
+                result = f'"{folder}" folder not found'
+
+        else:
+            result = 'No update needed for folder'
```

Comment:
another cool technique to remove what we have here (arrow code) is to invert some of the if statements. That will help to flatten the arrow code and make it easier to read and comprehend.

Read a bit about it https://blog.codinghorror.com/flattening-arrow-code/

Reply:
Thanks for the link, I hadn't heard of arrow code before. I've flattened it, but I fear it's now crossed over into the 'too clever' side. Thoughts?

Reply:
I like it. It's a good opportunity to talk about the "short circuit" pattern (not sure if that is the documented name) as well. Instead of having an else you return from the first branch.

```py
        #: Default result to report; update as needed
        result = 'No update needed for folder'

        if self.item_report['folder_fix'] == 'Y':
            folder = self.item_report['folder_new']
            move_result = self.item.move(folder)

            if not move_result:
                #: .move() returns false/empty if folder not found
                result = f'"{folder}" folder not found'
            else:
                result = f'Failed to move item to "{folder}" folder'
                if move_result['success']:
                    result = f'Item moved to "{folder}" folder'

        self.item_report['folder_result'] = result
```

can become

```py
        #: Default result to report; update as needed
        result = 'No update needed for folder'

        if self.item_report['folder_fix'] == 'Y':
            folder = self.item_report['folder_new']
            move_result = self.item.move(folder)

            if not move_result:
                #: .move() returns false/empty if folder not found
                self.item_report = f'"{folder}" folder not found'

				return
            
            result = f'Failed to move item to "{folder}" folder'
            if move_result['success']:
                result = f'Item moved to "{folder}" folder'

        self.item_report['folder_result'] = result
```

This removes another else branch.

I would probably write this like...

```py
        #: Default result to report; update as needed
        result = 'No update needed for folder'

        if self.item_report['folder_fix'].lower() != 'y':
            self.item_report['folder_result'] = 'No update needed for folder'

            return
        
        folder = self.item_report['folder_new']
        move_result = self.item.move(folder)

        if not move_result:
            #: .move() returns false/empty if folder not found
            self.item_report['folder_result'] = f'"{folder}" folder not found'

            return
     
        if not move_result['success']:
            self.item_report['folder_result'] = f'Failed to move item to "{folder}" folder'

            return

        self.item_report['folder_result'] = f'Item moved to "{folder}" folder'
```

The benefits in my opinion are 

1. no nesting
1. if statements short circuit the body of the function acting as guards
1. success is at the bottom of the function meaning all guard tests have passed

Reply:
Thanks for the example. I think I'm falling into the same trap as one of the blog commenters- for some reason, I always want a single return, and that isn't really necessary. I'll dig into this tomorrow.

Reply:
One of the big sells to convert me on this way back was that I didn't have to read, comprehend, and understand the entire function when debugging -- unless it was successful. The savings grow as the functions get more complex .

Reply:
Question as I flatten out my fixers: in your opinion, am I not being explicit enough here?

```py
update_result = self.item.update(update_dict)
#:item.update() returns False if it fails
if not update_result:
    self.item_report['tags_title_result'] = f'Failed to update {", ".join(messages)}'
    return
```

vs.

```py
if not self.item.update(update_dict):
    self.item_report['tags_title_result'] = f'Failed to update {", ".join(messages)}'
    return
```

Reply:
I would mostly choose option 1 as it is simpler to debug. 

I do like a new lines above returns but I will focus on your original question :)

Reply:
Ok, it's all flattened out. I've thought about it and decided to stick with the return statements as they are, because it visually groups them with the case they follow and separates the individual cases.